Add Salt Bundle functionality extension with PIP modules (#1674)
uyuni-project/uyuni-docs: modules/client-configuration/pages/contact-methods-saltbundle.adoc

This method is not recommended nor supported and exists for development purposes only.
Set [option]``web.ssh_use_salt_thin`` to `true` in the [path]``/etc/rhn/rhn.conf`` configuration file.
====

> == Extending Salt Bundle functionality with Python packages using pip

Reviewer: I'm sorry that I must start with this again.
```suggestion
== Extend Salt Bundle with Python packages using pip
```
In titles we prefer verbal form.  I think this is short for "How to extend the Salt Bundle ...".  The -ing form is reserved for procedure titles.  That's how Lana argued.  I'd like to stay with this style.  If you agree here, we also should change the .changelog entry.

> The Salt Bundle includes [literal]``pip`` to make it possible to extend the functionality of the bundled Salt Minion with extra Python packages.

Reviewer: ```suggestion
The Salt Bundle includes [literal]``pip`` to make it possible to add features to the Salt Bundle connection method with extra Python packages.
```
Your version can be read as if this is a possibility to add Python packages to the Client in general.

By all means, we must avoid the term "minion".  We prefer the term "client" here.

Last I'm wondering whether we should say "Python packages" or "Python modules" (pip has probably something to do with packages).

Author: No, now it's not correct. Salt Bundle is not a connection method by itself, it's more like a different flavour of Salt Minion.

Python package here is not the RPM or DEB package of the linux, but the package of Python (a set of modules). And we are talking here only about extending the bundled python with extra packages which will be available for the bundled python and salt bundle (bundled salt minion to be exact).

```suggestion
The Salt Bundle includes [literal]``pip`` to make it possible to extend the functionality of the Salt Bundle with extra Python packages.
```
But here I would like to avoid to mention Salt Bundle twice in the same sentance, so in original I've replaced the second occurance with the bundled Salt Minion as it's actually the same in this context.

Reviewer: thanks, I think I now understand it.  We can close this as resolved.

The default location to deploy modules on calling [command]``salt <minion_id> pip.install <module_name>`` is [path]``/var/lib/venv-salt-minion/local``.

[NOTE]
====

> In case if [path]``/var/lib/venv-salt-minion/local`` is not applicable for some reasons and the other location for the modules is required the destination path can be overridden by setting [literal]``VENV_PIP_TARGET`` environment variable for the [literal]``venv-salt-minion.service``.

Reviewer 2: ```suggestion
If needed, the path [path]``/var/lib/venv-salt-minion/local`` can be overridden by setting the [literal]``VENV_PIP_TARGET`` environment variable for the [literal]``venv-salt-minion.service``.
```

It is recommended to use systemd drop-in config file for the service. It could be done with the config file [path]``/etc/systemd/system/venv-salt-minion.service.d/10-pip-destination.conf``:
----
[Service]
Environment=VENV_PIP_TARGET=/new/path/local/venv-salt-minion/pip
----
====

[IMPORTANT]
====

> The Python packages installed through [literal]``pip`` are not changing on updating the Salt Bundle. To ensure that such packages are available and functional after an update, it is recommended to install them with a Salt state that is applied after Salt Bundle updates.

Author: It's not strange, but I have no idea about it. Thanks for pointing.

> The Salt Bundle does not maintain the modules installed with [literal]``PIP``. In case of updating the Salt Bundle it is possible that modules should be reinstalled, for example on updating the Salt Bundle with updated Python inside.

Reviewer 2: ```suggestion
The Salt Bundle does not keep Python packages installed through [literal]``pip`` on updates. To ensure that such packages are available after an update, it is recommended to install them with a Salt state that is applied after Salt Bundle updates.
```

Reviewer: It is probably not needed to explain the implementation too detailed.  I'd basically say:

> The Salt Bundle is self-contained and works independent from installed system components such as the Python interpreter and the Python modules.
> The Salt Bundle does neither update nor uninstall a system component such as Python.

I'm not sure whether we must explicitly say that updating Python will happen in the run of regular system updates.

Reviewer 2: The point here is that a user can install additional Python packages into the bundle, but those aren't updated when everything else in the bundle (Python, Salt, the Python packages we include) is updated. 

I previously thought that a Salt Bundle update would remove these additionally installed Python packages (which might still happen once we tackle https://confluence.suse.com/display/SUSEMANAGER/Problems+around+updating+Salt+Minion+%28bundled%29+with+Salt), but apparently that's not the case now. Either way, while these additional packages are placed into the bundle, it needs to be clear that the bundle is not actively doing anything with them.

Author: Is it sounds good if reword it this way?
```suggestion
The Python packages installed through [literal]``pip`` are not changing on updating the Salt Bundle. To ensure that such packages are available and functional after an update, it is recommended to install them with a Salt state that is applied after Salt Bundle updates.
```

Reviewer 2: Sounds good!
